streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/Graph.java

```diff
@@ -0,0 +1,348 @@
+                }
+            }
+        }
+    }
+
+    private void validateMinCostGraph() {
+        if (isResidualGraph) {
+            throw new IllegalStateException("Should not be residual graph to solve min cost flow");
+        }
+
+        /*
+         Check provided flow satisfying below constraints:
+         1. Input flow and output flow for each node should be the same except for source and destination node
+         2. Output flow of source and input flow of destination should be the same
+        */
+
+        final Map<V, Long> inFlow = new HashMap<>();
+        final Map<V, Long> outFlow = new HashMap<>();
+        populateInOutFlow(inFlow, outFlow);
+
+        for (final Entry<V, Long> in : inFlow.entrySet()) {
+            if (in.getKey().equals(sourceNode) || in.getKey().equals(sinkNode)) {
+                continue;
+            }
+            final Long out = outFlow.get(in.getKey());
+            if (!Objects.equals(in.getValue(), out)) {
+                throw new IllegalStateException("Input flow for node " + in.getKey() + " is " +
+                    in.getValue() + " which doesn't match output flow " + out);
+            }
+        }
+
+        final Long sourceOutput = outFlow.get(sourceNode);
+        final Long sinkInput = inFlow.get(sinkNode);
+        if (!Objects.equals(sourceOutput, sinkInput)) {
+            throw new IllegalStateException("Output flow for source " + sourceNode + " is " + sourceOutput
+                + " which doesn't match input flow " + sinkInput + " for sink " + sinkNode);
+        }
+    }
+
+    private void cancelNegativeCycles() {
+        if (!isResidualGraph) {
+            throw new IllegalStateException("Should be residual graph to cancel negative cycles");
+        }
+        boolean cyclePossible = true;
+        while (cyclePossible) {
+            cyclePossible = false;
+            for (final V node : nodes) {
+                final Map<V, V> parentNodes = new HashMap<>();
+                final Map<V, Edge> parentEdges = new HashMap<>();
+                final V nodeInCycle = detectNegativeCycles(node, parentNodes, parentEdges);
+                if (nodeInCycle == null) {
+                    continue;
+                }
+                cyclePossible = true;
+                V parentNode = parentNodes.get(nodeInCycle);
+                Edge parentEdge = parentEdges.get(nodeInCycle);
+
+                // Find max possible negative flow
+                int possibleFlow = parentEdge.residualFlow;
+                for (V curNode = parentNode; curNode != nodeInCycle; curNode = parentNodes.get(curNode)) {
+                    parentEdge = parentEdges.get(curNode);
+                    possibleFlow = Math.min(possibleFlow, parentEdge.residualFlow);
+                }
+
+                // Update graph by removing negative flow
+                parentNode = parentNodes.get(nodeInCycle);
+                parentEdge = parentEdges.get(nodeInCycle);
```

Comment:
Why do we need this update? Seems it would not change? Can we remove this line?

(We cannot make `parentEdge` `final` thought it seems, as it's set to a different value below)

Reply:
Actually `parentEdge` is changing from line 283